assets/shiny-server.js

```diff
@@ -273,10 +317,85 @@
           debug("NOT opening channel " + channel.id);
         }
       }
+      
+      // Send any buffered messages.
+      var msg;
+      while ((msg = self._buffer.shift())){
+        self._conn.send(msg);
+      }
     };
-    this._conn.onclose = function(e) {
+
+    // @param reconnect If true, show the reconnecting dialog. If false, show 
+    // disconnected
+    this.startDisconnect = function(){
+      self._disconnectDialog = true;
+      $('body').addClass('reconnecting');
+
+      var timeout = 15;
+      var reconnectingContent = '<button type="button" id="ss-reconnect-btn" class="ss-dialog-button">Reconnect ('+timeout+')</button><span class="ss-dialog-text">Trouble connecting to server</span>';
+      $('<div id="ss-connect-dialog">'+reconnectingContent+'<div class="ss-clearfix"></div></div><div id="ss-gray-out"></div>').appendTo('body');
+      $('#ss-reconnect-btn').click(function(){
+        $('ss-reconnect-btn').prop('disabled', true);
+        log("Attempting to reopen.");
+        self._openConnection();
+      });
+
+      var countdown = setInterval(function(){
+        timeout--;
+        updateDialog(timeout);
+      }, 1000);
+      updateDialog(timeout); //Update immediately
+
+      function updateDialog(timeout){
+        $('#ss-reconnect-btn').html('Reconnect (' + timeout + ')');
+        if (timeout <= 0 || self._disconnected){
+          clearInterval(countdown);
+          self._doClose();
+          $('body').removeClass('reconnecting');
+
+          $('#ss-connect-dialog').html('<button id="ss-reload-button" type="button" class="ss-dialog-button">Reload</button> Disconnected from the server.');
+          $('#ss-reload-button').click(function(){
+            location.reload();
+          });
+        }
+      }
+
+      self._clearReconnect = function(){
+        self._disconnectDialog = false;
+        clearInterval(countdown);
+        $('body').removeClass('reconnecting');
+        $('#ss-connect-dialog').remove();
+        $('#ss-gray-out').remove();
+        self._clearReconnect = function() {};
+      };
+    };
+
+    this.onConnClose = function(e) {
       log("Connection closed. Info: " + JSON.stringify(e));
       debug("SockJS connection closed");
+
+      if (self._disconnectDialog){
+        // This was a failed attempt to reconnect
+        alert(e.reason);
```

Comment:
I think this is also a candidate for a more graceful UI treatment than `alert`. I'll leave for a follow-on PR.